### PR TITLE
refactor: Voice Player date structure

### PR DESCRIPTION
**Pull request by HoonBaek**
## For Internal Contributors

[UIKIT-3263](https://sendbird.atlassian.net/browse/UIKIT-3263)

## Description Of Changes

* Use dux pattern for voice player data structure
* Add `IDLE` status to the VoicePlayerStatus
* Use document&element to refer the Audio element in the unmount scope of useEffect in useVoicePlayer

## Types Of Changes

What types of changes does your code introduce to this project?
Put an `x` in the boxes that apply_

- [x] Bugfix
- [ ] New feature
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance (ex) Prettier)
- [ ] Build configuration
- [x] Improvement (refactor code)

[UIKIT-3263]: https://sendbird.atlassian.net/browse/UIKIT-3263?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/hooks/VoicePlayer/dux/actionTypes.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export type InitializeAudioUnit = 'INITIALIZE_AUDIO_UNIT';
+export type SetCurrentPlayer = 'SET_CURRENT_PLAYER';
+export type OnVoicePlayerPlay = 'ON_VOICE_PLAYER_PLAY';
+export type OnVoicePlayerPause = 'ON_VOICE_PLAYER_PAUSE';
+export type OnCurrentTimeUpdate = 'ON_CURRENT_TIME_UPDATE';
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/dux/actionTypes.ts:5]: I added types, but I don't know it's a good idea

File: src/hooks/VoicePlayer/dux/initialState.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  currentPlayer: null | HTMLAudioElement;
+  currentGroupKey: string;
+  audioStorage: Record<GroupKey, AudioStorageUnit>;
+}
+
+export const voicePlayerInitialState: VoicePlayerInitialState = {
+  currentPlayer: null,
+  currentGroupKey: '',
+  audioStorage: {},
+};
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/dux/initialState.ts:34]: So initialState will look like
```javascript
const initialState = {
  currentPlayer: HTMLAudioElement,
  currentGroupKey: string (channelUrl + messageId),
  audioStorage: {
    [groupKey]: {
      playingStatus: VoicePlayerStatus,
      audioFile: File,
      playbackTime: number,
      duration: number,
    },
  },
}
```

File: src/hooks/VoicePlayer/dux/reducer.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+type OnVoicePlayerPlayPayload = { groupKey: string, audioFile: File };
+type OnVoicePlayerPausePayload = { groupKey: string };
+type OnCurrentTimeUpdatePayload = { groupKey: string };
+type PayloadType = (
+  InitializeAudioUnitPayload
+  | SetCurrentPlayerPayload
+  | OnVoicePlayerPlayPayload
+  | OnVoicePlayerPausePayload
+  | OnCurrentTimeUpdatePayload
+);
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/dux/reducer.ts:26]: I'm not sure if this is a good idea, here too

File: src/hooks/VoicePlayer/index.tsx [hunk 4]
```diff
@@ -1,129 +1,168 @@
...
+  audioFile?: File;
+  audioFileUrl?: string;
 }
 export interface VoicePlayerContext {
   play: (props: VoicePlayerPlayProps) => void;
-  stop: (groupKey?: string) => void;
-  addEventHandler: (props: VoicePlayerEvent) => void;
-  removeEventHandler: (groupKey: string, handlerId: string) => void;
+  pause: (groupKey?: string) => void;
+  stop: (text?: string) => void;
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/index.tsx:35]: pause needs specific groupKey which is made of channelUrl and messageId
but stop can called any string which contains channelurl or messageId or empty string

File: src/hooks/VoicePlayer/index.tsx [hunk 5]
```diff
@@ -1,129 +1,168 @@
...
-        playbackTime: audioPlayer?.currentTime,
-        duration: audioPlayer?.duration,
+    // Clear the previous AudioPlayer element
+    const voicePlayerRoot = document.getElementById(VOICE_PLAYER_ROOT_ID);
+    const voicePlayerAudioElement = document.getElementById(VOICE_PLAYER_AUDIO_ID);
+    if (voicePlayerAudioElement) {
+      voicePlayerRoot.removeChild(voicePlayerAudioElement);
+    }
+
+    new Promise((resolve) => {
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/index.tsx:91]: Here, I'm getting audioFile from three ways
1. already saved in the audioStorage
2. audioFile, the parameter of `play` (this is used in VoiceMessageInput)
3. audioFileUrl, the parameter of `play` (this is used in VoiceMessageItemBody)

File: src/utils/consts.ts [hunk 6]
```diff
@@ -8,6 +8,13 @@ export const VOICE_RECORDER_AUDIO_BITS = 128000;
 export const VOICE_PLAYER_PLAYBACK_BUFFER = 0.01;
 export const VOICE_PLAYER_DURATION_MIN_SIZE = 1000;
 
+/**
+ * Append Audio element to the root
+ * because I can't get the Audio element in the useEffect unmount scope
+ */
+export const VOICE_PLAYER_ROOT_ID = 'sendbirid-voice-player-provider-root';
```
Comment (sravan-s) [inline src/utils/consts.ts:15]: sendbirid -> sendbird

Comment (HoonBaek) [reply src/utils/consts.ts]: done! thanks

File: src/hooks/VoicePlayer/dux/actionTypes.ts (hunk 1 above)
Comment (sravan-s) [reply src/hooks/VoicePlayer/dux/actionTypes.ts:5]: whynot use the uppercase name as the const type? example -> `INITIALIZE_AUDIO_UNIT`

Comment (sravan-s) [reply src/hooks/VoicePlayer/dux/actionTypes.ts:5]: This is fine for now~ dont need to change

File: src/hooks/VoicePlayer/dux/initialState.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+export const VoicePlayerStatus = {
+  IDLE: 'IDLE',
+  PREPARING: 'PREPARING',
+  PLAYING: 'PLAYING',
+  PAUSED: 'PAUSED',
+  COMPLETED: 'COMPLETED',
+} as const;
+export type VoicePlayerStatus = typeof VoicePlayerStatus[keyof typeof VoicePlayerStatus];
+
+export type GroupKey = string;
```
Comment (sravan-s) [inline src/hooks/VoicePlayer/dux/initialState.ts:10]: Can you explain here what are the components of a group key?
comment is enough

Comment (HoonBaek) [reply src/hooks/VoicePlayer/dux/initialState.ts:10]: Ok I will add

Comment (HoonBaek) [reply src/hooks/VoicePlayer/dux/initialState.ts]: I move this type definition to the place where the generateGroupKey function is

File: src/hooks/VoicePlayer/dux/actionTypes.ts (hunk 1 above)
Comment (sravan-s) [reply src/hooks/VoicePlayer/dux/actionTypes.ts:5]: https://github.com/sendbird/sendbird-uikit-react/pull/398/files#diff-c23f1837a7a5060b1a091e4fa42ddd54f782ad6f9f563e79220549107ad7357c > use this pattern, its cleaner ~

File: src/hooks/VoicePlayer/dux/reducer.ts (hunk 3 above)
Comment (sravan-s) [reply src/hooks/VoicePlayer/dux/reducer.ts:26]: This can be made cleaner ~ keep this for now.. I will suggest after you release

**Review by sravan-s — APPROVED** [review]

File: src/hooks/VoicePlayer/index.tsx [hunk 8]
```diff
@@ -1,129 +1,168 @@
...
+          const audioFile = new File([blob], VOICE_MESSAGE_FILE_NAME, {
+            lastModified: new Date().getTime(),
+            type: VOICE_MESSAGE_MIME_TYPE,
+          });
+          resolve(audioFile);
+        });
+    }).then((audioFile: File) => {
+      const currentAudioUnit = audioStorage[groupKey] || AudioUnitDefaultValue() as AudioStorageUnit;
+      const audioPlayer = new Audio(URL?.createObjectURL?.(audioFile));
+      audioPlayer.id = VOICE_PLAYER_AUDIO_ID;
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/index.tsx:116]: Set `id` of Audio element

File: src/hooks/VoicePlayer/index.tsx [hunk 9]
```diff
@@ -1,129 +1,168 @@
...
+        voicePlayerDispatcher({
+          type: ON_CURRENT_TIME_UPDATE,
+          payload: {
+            groupKey,
+          },
+        });
+      };
+      audioPlayer?.play();
+      const voicePlayerRoot = document.getElementById(VOICE_PLAYER_ROOT_ID);
+      voicePlayerRoot.appendChild(audioPlayer);
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/index.tsx:147]: And append the Audio element to the root element

File: src/hooks/VoicePlayer/index.tsx [hunk 10]
```diff
@@ -1,129 +1,168 @@
...
   return (
     <VoicePlayerContext.Provider value={{
       play,
+      pause,
       stop,
-      addEventHandler,
-      removeEventHandler,
+      voicePlayerStore,
     }}>
+      <div id={VOICE_PLAYER_ROOT_ID} style={{ display: 'none' }} />
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/index.tsx:165]: This is the root element

File: src/hooks/VoicePlayer/useVoicePlayer.tsx [hunk 11]
```diff
@@ -1,109 +1,72 @@
...
 
+  const stopVoicePlayer = (text = '') => {
+    stop?.(text);
+  };
+
+  useEffect(() => {
+    return () => {
+      // Can't get the current AudioPlayer through the React hooks(useReducer or useState) in this scope
+      const voiceAudioPlayerElement = document.getElementById(VOICE_PLAYER_AUDIO_ID);
+      (voiceAudioPlayerElement as HTMLAudioElement)?.pause?.();
```
Comment (HoonBaek) [inline src/hooks/VoicePlayer/useVoicePlayer.tsx:59]: And get and use the Audio element to call `pause` here

**Pull request merged**